packages/ember-glimmer/tests/utils/shared-conditional-tests.js

```diff
@@ -374,11 +384,56 @@ export class SharedConditionalsTest extends AbstractConditionalsTest {
     this.assertText('T1F2');
   }
 
+  ['@test it maintains DOM stability when condition changes from a truthy to a different truthy value']() {
+    this.renderValues(true);
+
+    this.assertText('T1');
+
+    this.takeSnapshot();
+
+    this.runTask(() => set(this.context, 'cond1', 'hello'));
+
+    this.assertText('T1');
+
+    this.assertInvariants();
+  }
+
+  ['@test it maintains DOM stability when condition changes from a falsy to a different falsy value']() {
+    this.renderValues(false);
+
+    this.assertText('F1');
+
+    this.takeSnapshot();
+
+    this.runTask(() => set(this.context, 'cond1', ''));
+
+    this.assertText('F1');
+
+    this.assertInvariants();
+  }
+}
+
+export class SharedHelperConditionalsTest extends SharedConditionalsTest {
+  renderValues(...values) {
+    let templates = [];
+    let context = {};
+
+    for (let i = 1; i <= values.length; i++) {
+      templates.push(this.templateFor({ cond: `cond${i}`, truthy: `t${i}`, falsy: `f${i}` }));
```

Comment:
I see. So the problem is that we got a little too greedy in the original test case. It was originally written as `{{#if cond1}}T1{{else}}F1{{/if}}`, but then we realized that it's generally important to smoke test that both templates ("default" and "inverse") have access to the ambient scope, so we changed it to be `{{#if cond1}}{{t}}1{{else}}{{f}}1{{/if}}` to test that it has access to `t` and `f` on the outer scope. (We also somehow decided that it is important to test that `{{t}}1` works, as opposed to just `{{t1}}`.)

So... we can do one of two things here:

1. We decide that "obviously" that the condition position and the value positions share the same sccope, and there are no risk of scope switching here, and since the condition is already dynamic (tests accessing the scope), there is no need of testing that in the value positions as well. In which case, we can generate something like `{{if cond1 "T1" "F1"}}`.

2. We maintain the spirit of the original test, and generates `{{if cond1 (concat t "1") (concat f "1")}}`.

3. We go above and beyond and make all three parts dynamic, like you did here.

I think all of these positions are actually quite reasonable, and perhaps it even makes some sense to want to test *all* of them. The problem I see now is that we decided to go with (3) here, but in the actually `inline-if-test.js` we *also* felt the need to have *some* tests for (1) (that's what "static argument", "it can use falsy param as binding", "it respects isTruthy when object changes" etc are about, if I understand correctly).

If we decided to care about multiple of those  scenarios, I think it makes more sense to use the same set of test cases and subclass multiple times with different `renderValues` implementations instead of doing one set of them here and the rest in an ad-hoc fashion in the concrete subclass.

For example, we have a pretty exhaustive sets of `isTruthy` tests in the shared test cases (bare object with `isTruthy`, `Ember.Object` various proxies, testing both true and false as the starting value, etc), but we only have a single row of that matrix in the concrete subclass for (1) at the moment:

```javascript
  ['@test it respects isTruthy when object changes']() {
     this.render(`{{if condition 'truthy' 'falsy'}}`, { condition: EmberObject.create({ isTruthy: false }) });
 
     this.assertText('falsy');
 
     this.runTask(() => this.rerender());
 
     this.assertText('falsy');
 
     this.runTask(() => set(this.context, 'condition', EmberObject.create({ isTruthy: true })));
 
     this.assertText('truthy');
 
     this.runTask(() => set(this.context, 'condition', EmberObject.create({ isTruthy: false })));
 
     this.assertText('falsy');
   }
```

The TL;DR is that I think we should either decide that one of these variations is enough, and remove most of the ad-hoc tests in the subclass, or we subclass this multiple times.

Does that make sense?